contracts/zerodev/account/kernel/WalletKernel.sol

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@gnosis.pm/safe-contracts/contracts/GnosisSafe.sol";
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+import "../../plugin/IPlugin.sol";
+import "../../../core/Helpers.sol";
+import "../../../interfaces/IAccount.sol";
+import "../../utils/OpinionatedExec.sol";
+import "./Compatibility.sol";
+
+struct WalletKernelStorage {
+    address owner;
+    uint256 nonce;
+}
+
+
+/// @title Wallet Kernel
+/// @author taek<leekt216@gmail.com>
+/// @notice wallet kernel for minimal wallet functionality
+/// @dev supports only 1 owner and 1 threshold, multiple plugins
+contract WalletKernel is IAccount, EIP712, Compatibility {
+
+    function getWalletKernelStorage() internal pure returns (WalletKernelStorage storage ws) {
+        bytes32 storagePosition = bytes32(uint256(keccak256("zero-dev.account.walletkernel")) - 1);
+        assembly {
+            ws.slot := storagePosition
+        }
+    }
+
+    uint256 constant private SIG_VALIDATION_FAILED = 1;
+
+    address public immutable entryPoint;
+
+    error QueryResult(bytes result);
+
+    constructor(address _entryPoint) EIP712("WalletKernel", "0.0.1") {
+        entryPoint = _entryPoint;
+    }
+
+    /// @notice Query plugin for data
+    /// @dev this function will always fail, it should be used only to query plugin for data using error message
+    /// @param _plugin Plugin address
+    /// @param _data Data to query
+    function queryPlugin(address _plugin, bytes calldata _data) external {
+        (bool success, bytes memory _ret) = OpinionatedExec.delegateCall(_plugin, _data);
+        if(success) {
+            revert QueryResult(_ret);
+        } else {
+            assembly {
+                revert(add(_ret, 32), mload(_ret))
+            }
+        }
+    }
+
+
+    /**
+     * Execute a call but also revert if the execution fails.
+     * The default behavior of the Safe is to not revert if the call fails,
+     * which is challenging for integrating with ERC4337 because then the
+     * EntryPoint wouldn't know to emit the UserOperationRevertReason event,
+     * which the frontend/client uses to capture the reason for the failure.
+     */
+    function executeAndRevert(
+        address to,
+        uint256 value,
+        bytes calldata data,
+        Enum.Operation operation
+    ) external {
+        require(msg.sender == entryPoint, "account: not from entrypoint");
```

Comment:
Do you think we should have a way for users to call this account independently of the EntryPoint?  Cuz if say something goes wrong with the EntryPoint/4337, we don't want these accounts to be completely inaccessible.

Reply:
that makes sense i will change this to be "onlyOwnerOrEntryPoint"